# HG changeset patch
# Parent  28701ca70422b882028b7fd0e6c52401b394041a
Fix build on FreeBSD 10.3

- properly use _XOPEN_SOURCE=700 instead of _POSIX_C_SOURCE;
- fix missing includes;
- add compat file for endian.h (maybe it should be the other way or
  maybe I should just improve the CMake "finder").

diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -55,14 +55,20 @@
 SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wextra -Wall -Wstrict-prototypes -std=c99")
 SET(CMAKE_C_FLAGS_${CMAKE_BUILD_TYPE} "")
 
+IF (CMAKE_SYSTEM_NAME MATCHES "Linux")
+    ADD_DEFINITIONS(
+        # see feature_test_macros(7)
+        "-D_XOPEN_SOURCE=700"
+        "-D_BSD_SOURCE=1"
+        "-D_DEFAULT_SOURCE=1"
+    )
+ENDIF ()
+
+IF (APPLE)
+    ADD_DEFINITIONS("-D_DARWIN_C_SOURCE=1")
+ENDIF ()
+
 ADD_DEFINITIONS(
-    # Only relevant for the GNU libc:
-    "-D_POSIX_C_SOURCE=200809L"
-    "-D_BSD_SOURCE=1"
-    "-D_DEFAULT_SOURCE=1"
-
-    "-D_DARWIN_C_SOURCE=1"
-
     "-DLGTD_BIG_ENDIAN_SYSTEM=${BIG_ENDIAN_SYSTEM}"
     "-DLGTD_SIZEOF_VOID_P=${CMAKE_SIZEOF_VOID_P}"
 
diff --git a/compat/FreeBSD/endian.h b/compat/FreeBSD/endian.h
new file mode 100644
--- /dev/null
+++ b/compat/FreeBSD/endian.h
@@ -0,0 +1,3 @@
+#pragma once
+
+#include <sys/endian.h>
diff --git a/core/listen.c b/core/listen.c
--- a/core/listen.c
+++ b/core/listen.c
@@ -19,6 +19,7 @@
 #include <sys/socket.h>
 #include <sys/stat.h>
 #include <sys/un.h>
+#include <netinet/in.h>
 #include <assert.h>
 #include <err.h>
 #include <errno.h>
diff --git a/lifx/broadcast.c b/lifx/broadcast.c
--- a/lifx/broadcast.c
+++ b/lifx/broadcast.c
@@ -18,6 +18,7 @@
 #include <sys/queue.h>
 #include <sys/tree.h>
 #include <arpa/inet.h>
+#include <netinet/in.h>
 #include <assert.h>
 #include <endian.h>
 #include <err.h>
diff --git a/lifx/gateway.c b/lifx/gateway.c
--- a/lifx/gateway.c
+++ b/lifx/gateway.c
@@ -17,6 +17,7 @@
 
 #include <sys/queue.h>
 #include <sys/tree.h>
+#include <netinet/in.h>
 #include <assert.h>
 #include <endian.h>
 #include <err.h>
diff --git a/tests/core/pipe/test_pipe_open_fifo_already_exists.c b/tests/core/pipe/test_pipe_open_fifo_already_exists.c
--- a/tests/core/pipe/test_pipe_open_fifo_already_exists.c
+++ b/tests/core/pipe/test_pipe_open_fifo_already_exists.c
@@ -1,6 +1,7 @@
 #include "pipe.c"
 
 #include <sys/tree.h>
+#include <sys/stat.h>
 #include <endian.h>
 #include <limits.h>
 
